File: app/src/main/java/com/awscherb/cardkeeper/ui/base/BaseAdapter.java
package com.awscherb.cardkeeper.ui.base;

import android.support.v7.widget.RecyclerView;

import com.awscherb.cardkeeper.data.model.BaseModel;

import java.util.List;

public abstract class BaseAdapter<T extends BaseModel, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> {

    private List<T> objects;

    public BaseAdapter(List<T> objects) {
        this.objects = objects;
    }

    @Override
    public void onBindViewHolder(VH holder, int position) {
        onBindViewHolder(holder, getItem(position));
    }

    public abstract void onBindViewHolder(VH holder, T item);

    public T getItem(int position) {
        return objects.get(position);
    }

    @Override
    public int getItemCount() {
        return objects.size();
    }

    public void swapObjects(List<T> newObjects) {
        objects = newObjects;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/awscherb/cardkeeper/ui/base/BaseAdapter.kt
package com.awscherb.cardkeeper.ui.base

import androidx.recyclerview.widget.RecyclerView

import com.awscherb.cardkeeper.data.model.BaseModel

abstract class BaseAdapter<T> constructor(
    private var objects: List<T> = ArrayList()
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() where T : BaseModel {

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        onBindViewHolder(holder, this[position])
    }

    abstract fun onBindViewHolder(holder: RecyclerView.ViewHolder, item: T)

    operator fun get(position: Int): T = objects[position]

    override fun getItemCount() = objects.size

    fun swapObjects(newObjects: List<T>) {
        objects = newObjects
        notifyDataSetChanged()
    }

}
